scripts/appbuilder

```diff
@@ -212,34 +212,32 @@ get_image()
 # export a flattened copy of the container
 export_image()
 {
-    TEMP_FILE=$(mktemp)
-    TEMP_DIR=$(mktemp -d)
     APPENV_FILENAME="appenv.json"
-
-    # note that we have to start up the image to flatten it
-    TEMP_INAME=$(docker run -d $IMAGE_NAME)
-    TEMP_APPENV=$APPENV_FILENAME-$TEMP_INAME
+    TEMP_APPENV=$APPENV_FILENAME-$(date +%s)
     
     docker image inspect --format='{{json .Config}}' $IMAGE_NAME > $TEMP_APPENV
+    LOWER_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.LowerDir}}' $IMAGE_NAME | tr -d '"')
+    UPPER_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.UpperDir}}' $IMAGE_NAME | tr -d '"')
+    WORK_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.WorkDir}}' $IMAGE_NAME | tr -d '"')
 
-    docker stop $TEMP_INAME >/dev/null
-    docker export $TEMP_INAME -o $TEMP_FILE
-    docker rm $TEMP_INAME >/dev/null
-    if [ -z $DELETE_IMAGE ]; then
-        docker rmi -f $IMAGE_NAME
-    fi
+    rm -rf $OUTPUT_NAME
 
-    # create the 'appdir' from the docker export
-    tar xf $TEMP_FILE -C $TEMP_DIR
-    mv $TEMP_APPENV $TEMP_DIR/$APPENV_FILENAME
-    rm -rf $OUTPUT_NAME $TEMP_FILE
+    MNT_PNT_DIR=$(mktemp -d)
 
     if [ "$OUTPUT_FORMAT" == "dir" ]; then
-        mv $TEMP_DIR $OUTPUT_NAME
+        EXTRACT_CMD="cp -r $MNT_PNT_DIR $OUTPUT_NAME"
     elif [ "$OUTPUT_FORMAT" == "cpio" ]; then
-        myst mkcpio $TEMP_DIR $OUTPUT_NAME
+        EXTRACT_CMD="myst mkcpio $MNT_PNT_DIR $OUTPUT_NAME"
     fi
-    rm -rf $TEMP_DIR
+
+    sudo mount -t overlay overlay -olowerdir=$LOWER_DIR,upperdir=$UPPER_DIR,workdir=$WORK_DIR $MNT_PNT_DIR;
+    sudo $EXTRACT_CMD;
+    sudo umount $MNT_PNT_DIR;
+
+    # change ownership back to current effective user and group
+    sudo chown -R $USER:$(id -ng) $OUTPUT_NAME
```

Comment:
If we change owner here, the CPIO archive will have the wrong owners. Better do it on the temp_dir before it is copied or compressed.

Reply:
Oh you mean the permissions for the individual files within the archive?